.github/workflows/publish-and-release.yaml

```diff
@@ -0,0 +1,136 @@
+
+on:
+  push:
+
+jobs:
+  repo-prep:
+    name: Prepare versioning and changelogs for release
+    runs-on: ubuntu-latest
+    outputs:
+      changelog-body: ${{ steps.generate-changelog.outputs.stdout }}
+      latest-tag: ${{ steps.get-latest-tag.outputs.tag }}
+      version: ${{ steps.set-version.outputs.version }}
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v3
+      - name: Get latest existing tag
+        uses: actions-ecosystem/action-get-latest-tag@v1
+        id: get-latest-tag
+        with:
+          semver_only: true
+      - name: Set new version
+        uses: paulhatch/semantic-version@v4.0.2
+        id: set-version
+        with:
+          tag_prefix: "v"
+          format: "${major}.${minor}.${patch}"
+          major_pattern: "(MAJOR)"
+          minor_pattern: "(MINOR)"
+      - name: Generate changelog since last tag
+        uses: mathiasvr/command-output@v1
+        id: generate-changelog
+        with:
+          run: git log --format="* %s" ${{ steps.get-latest-tag.outputs.tag }}..HEAD | { grep -v "(IGNORE)" || :; }
+      - name: Log version & changelog
+        run: |
+          echo "Version: $VERSION"
+          echo "Version tag: $VERSION_TAG"
+          echo "Latest tag detected: $LATEST_TAG"
+          echo "Changelog: $CHANGELOG"
+        env:
+          VERSION: ${{ steps.set-version.outputs.version }}
+          VERSION_TAG: ${{ steps.set-version.outputs.version_tag }}
+          LATEST_TAG: ${{ steps.get-latest-tag.outputs.tag }}
+          CHANGELOG: ${{ steps.changelog.outputs.stdout }}
+      # - name: Prevent empty release
+      #   if: ${{ steps.changelog.outputs.stdout == '' }}
+      #   uses: actions/github-script@v3
+      #   with:
+      #     script: |
+      #       core.setFailed("No changes since prior release")
+      - name: Update changelog
+        run: |
+          (head -8 CHANGELOG.md && echo "## $VERSION" && date "+_%e %B %Y_" && echo "" && (echo "$CHANGELOG" | sed -E 's_\(#([0-9]+)\)_([#\1](https://github.com/contentauth/c2pa-rs/pull/\1)\)_') && echo "" && tail -n +9 CHANGELOG.md) > CHANGELOG.new.md
+          mv CHANGELOG.new.md CHANGELOG.md
+        env:
+          VERSION: ${{ steps.set-version.outputs.version }}
+          CHANGELOG: ${{ steps.changelog.outputs.stdout }}
+  release-crate:
+    name: Release c2patool Rust crate
+    needs: repo-prep
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v3
+      - name: Install Rust toolchain
+        uses: actions-rs/toolchain@v1
+        with:
+          toolchain: stable
+      - name: Bump crate versions
+        run: |
+          sed -i "s/^version = \"[^\"]*\"$/version = \"$VERSION\"/;" Cargo.toml
+        env:
+          VERSION: ${{ needs.repo-prep.outputs.version }}
+      - name: Report differences for "prepare (release)" commit
+        run: git diff
+      - name: Commit Cargo.toml and changelog
```

Comment:
The steps up through commit cargo.toml and changelog probably need to move to `repo-prep` job; otherwise, the binaries generated in `publish-binaries` may or may not self-identify with the new version. I think this would be a race condition on whether this part of this job executed before or after `publish-binaries` for each platform.

Reply:
which steps specifically?

Reply:
All of them from the start of the job through "Commit Cargo.toml ..."

Reply:
Actually, the "checkout repo" should not move. It would be redundant in the other job and remains necessary in this one.